src/epanet.c

```diff
@@ -2794,6 +2743,72 @@ int DLLEXPORT EN_setdemandmodel(EN_Project p, int model, double pmin,
     return 0;
 }
 
+int DLLEXPORT EN_adddemand(EN_Project p, int index, double dmnd,
+                           char *dmndpat, const char *dmndname, int *dmndkey)
+/*----------------------------------------------------------------
+**  Input:   index = node index
+**           dmnd = base demand value
+**           dmndpat = demand pattern name
+**           dmndname = name of demand category
+**  Output:  dmndkey = key value of new demand category
+**  Returns: error code
+**  Purpose: adds a new demand category to a junction's demand list
+**----------------------------------------------------------------
+*/
+{
+    Network *net = &p->network;
+
+    int patIndex, error = 0;
+
+    *dmndkey = -1;
+
+    // Check that node exists
+    if (!p->Openflag) return 102;
+    if (index <= 0 || index > net->Nnodes) return 203;
+
+    // Do nothing if node is not a junction
+    if (index > net->Njuncs) return 0;
+
+    // Check that demand pattern exists
+    if (dmndpat == NULL || strlen(dmndpat) == 0) patIndex = 0;
+    else
+    {
+        error = EN_getpatternindex(p, dmndpat, &patIndex);
+        if (error != 0) return error;
+    }
+
+    // Append a new demand to the node's demand list
+    *dmndkey = add_demand(net->Node[index].D, dmnd / p->Ucf[FLOW], patIndex, dmndname);
+    return 0;
+}
+
+int DLLEXPORT EN_deletedemand(EN_Project p, int nodeIndex, int demandIndex)
```

Comment:
So when you add the demand you get a key and when you delete it you use an index? Seems confusing.

Reply:
Yes it is. Again, I don't understand how keys would work given the current code (either yours or mine) since they aren't stored anywhere. I suppose a toolkit client could write code to do so, but that would be quite a burden. And since access to list data by either index or key is done by simply traversing the list, there is no operational difference I see between the two of them.